src/gpu/intel/ocl/stream.cpp

```diff
@@ -136,6 +136,91 @@ status_t stream_t::barrier() {
     return impl()->barrier();
 }
 
+#ifdef DNNL_EXPERIMENTAL_ASYNC_VERBOSE
+
+status_t stream_t::register_async_tracker(cl_event event) {
+
+    // Check if there is any event in the queue - for reference
+    // implementation, there is no kernel implementation on queue to
+    // record for.
+    if (!event) return status::success;
+
+    // queue the event during primitive execution
+    cl_event exec_evt = event;
+    cl_int retain_status = clRetainEvent(exec_evt);
+    if (retain_status != CL_SUCCESS) return status::runtime_error;
+    async_tracked_event_ = exec_evt;
+
+    async_timing_data_ = new async_timing_data_t;
+    async_timing_data_->profiler_enabled = is_profiling_enabled();
+
+    cl_int err = clSetEventCallback(
+            exec_evt, CL_COMPLETE, async_tracker_callback, async_timing_data_);
+
+    return (err == CL_SUCCESS) ? status::success : status::runtime_error;
+}
+
+status_t stream_t::get_async_exec_times(
+        double *start_ms, double *end_ms, double *duration_ms) {
+
+    if (!async_tracked_event_) {
+        // if there is no tracked event, we fall back to using
+        // stream.wait() for capturing execution time.
+        wait();
+        *duration_ms = get_msec() - *start_ms;
+        delete async_timing_data_;
+        async_timing_data_ = nullptr;
+
+        return status::success;
+
+    } else {
+        cl_int err = clWaitForEvents(1, &async_tracked_event_);
```

Comment:
Isn't this call synchronous? We enqueue a kernel, record an event and execution blocks here, until the kernel finishes. Am I missing something?